File: windows/QMK Toolbox/Usb/Bootloader/CaterinaDevice.cs
using System.Threading.Tasks;

namespace QMK_Toolbox.Usb.Bootloader
{
    class CaterinaDevice : BootloaderDevice
    {
        public string ComPort { get; }

        public CaterinaDevice(UsbDevice d) : base(d)
        {
            Type = BootloaderType.Caterina;
            Name = "Caterina";
            PreferredDriver = "usbser";
            IsEepromFlashable = true;

            ComPort = FindComPort();
        }

        public async override Task Flash(string mcu, string file)
        {
            if (ComPort == null)
            {
                PrintMessage("COM port not found!", MessageType.Error);
                return;
            }

            await RunProcessAsync("avrdude.exe", $"-p {mcu} -c avr109 -U flash:w:\"{file}\":i -P {ComPort}");
        }

        public async override Task FlashEeprom(string mcu, string file)
        {
            if (ComPort == null)
            {
                PrintMessage("COM port not found!", MessageType.Error);
                return;
            }

            await RunProcessAsync("avrdude.exe", $"-p {mcu} -c avr109 -U eeprom:w:\"{file}\":i -P {ComPort}");
        }

        public override string ToString() => $"{base.ToString()} [{ComPort}]";
    }
}

File: linux/src/qmk_toolbox/bootloader/caterina_device.py
from .bootloader_device import BootloaderDevice
from .bootloader_type import BootloaderType


class CaterinaDevice(BootloaderDevice):
    def __init__(self, usb_device):
        super().__init__(usb_device)
        self.bootloader_type = BootloaderType.CATERINA
        self.name = "Caterina"
        self.preferred_driver = "cdc_acm"
        self.is_eeprom_flashable = True
    
    async def flash(self, mcu: str, file_path: str) -> int:
        port = self._find_port()
        if not port:
            from ..message_type import MessageType
            self._print_message("Could not find serial port for device", MessageType.ERROR)
            return 1
        
        args = ["-p", mcu, "-c", "avr109", "-U", f"flash:w:{file_path}:i", "-P", port]
        return await self._run_process("avrdude", args)
    
    async def flash_eeprom(self, mcu: str, file_path: str) -> int:
        port = self._find_port()
        if not port:
            from ..message_type import MessageType
            self._print_message("Could not find serial port for device", MessageType.ERROR)
            return 1
        
        args = ["-p", mcu, "-c", "avr109", "-U", f"eeprom:w:{file_path}:i", "-P", port]
        return await self._run_process("avrdude", args)
    
    def _find_port(self):
        import glob
        ports = glob.glob('/dev/ttyACM*') + glob.glob('/dev/ttyUSB*')
        return ports[0] if ports else None
